cupy/core/core.pyx

```diff
@@ -1853,26 +1854,26 @@ cdef str _cupy_header = ''.join(
 # This is indirect include header list.
 # These header files are subject to a hash key.
 cdef list _cupy_extra_header_list = [
-    'cupy/complex/complex.h',
```

Comment:
I think at least `cupy/complex/complex.h` needs to go first because it's the header of the headers...

Reply:
Sorry, don't mind me. I confused myself 😅

Reply:
Sorry again @kmaehashi, on a third thought we still need `complex.h` to be the first 😅 On CUDA, the extra header list is only used to be part of cache key, but on HIP we currently need to explicitly include all headers due to a bug in hipRTC:
https://github.com/cupy/cupy/blob/dadf7501b3111933f9f2a5295775f4b8511ecb53/cupy/cuda/compiler.py#L721-L730
If you could make the change I'll try testing this PR on HIP.

Reply:
I agree with @leofang here!

Reply:
I just pushed the change, @leofang can you verify if this works? thanks!

Reply:
Let me check

Reply:
No it doesn't work:
```
$ pytest tests/cupy_tests/core_tests/test_raw.py -x
========================================================================= test session starts =========================================================================
platform linux -- Python 3.7.8, pytest-6.0.2, py-1.9.0, pluggy-0.13.1
rootdir: /home/leofang/dev/cupy_rocm350, configfile: setup.cfg
collected 231 items                                                                                                                                                   

tests/cupy_tests/core_tests/test_raw.py F

============================================================================== FAILURES ===============================================================================
____________________________________________________ TestRaw_param_0_{backend='nvrtc', in_memory=False}.test_basic ____________________________________________________

self = <cupy.cuda.compiler._NVRTCProgram object at 0x7f4b9ea4d750>
options = ('-I/home/leofang/dev/cupy_rocm350/cupy/core/include', '-I/opt/rocm/include', '-arch=gfx906'), log_stream = None

    def compile(self, options=(), log_stream=None):
        try:
            if self.name_expressions:
                for ker in self.name_expressions:
                    nvrtc.addAddNameExpression(self.ptr, ker)
>           nvrtc.compileProgram(self.ptr, options)

cupy/cuda/compiler.py:605: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   cpdef compileProgram(intptr_t prog, options):

cupy_backends/cuda/libs/nvrtc.pyx:116: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   check_status(status)

cupy_backends/cuda/libs/nvrtc.pyx:128: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   raise NVRTCError(status)
E   cupy_backends.cuda.libs.nvrtc.NVRTCError: HIPRTC_ERROR_COMPILATION (6)

cupy_backends/cuda/libs/nvrtc.pyx:62: NVRTCError

During handling of the above exception, another exception occurred:

self = <<cupy_tests.core_tests.test_raw.TestRaw_param_0_{backend='nvrtc', in_memory=False} testMethod=test_basic>  parameter: {'backend': 'nvrtc', 'in_memory': False}>

    def test_basic(self):
>       x1, x2, y = self._helper(self.kern, cupy.float32)

tests/cupy_tests/core_tests/test_raw.py:453: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/cupy_tests/core_tests/test_raw.py:446: in _helper
    x1 = cupy.arange(N**2, dtype=dtype).reshape(N, N)
cupy/_creation/ranges.py:57: in arange
    _arange_ufunc(typ(start), typ(step), ret, dtype=dtype)
cupy/core/_kernel.pyx:1103: in cupy.core._kernel.ufunc.__call__
    kern = self._get_ufunc_kernel(dev_id, op, arginfos)
cupy/core/_kernel.pyx:1126: in cupy.core._kernel.ufunc._get_ufunc_kernel
    kern = _get_ufunc_kernel(
cupy/core/_kernel.pyx:900: in cupy.core._kernel._get_ufunc_kernel
    return _get_simple_elementwise_kernel(
cupy/core/_kernel.pyx:61: in cupy.core._kernel._get_simple_elementwise_kernel
    module = compile_with_cache(module_code, options)
cupy/core/core.pyx:2019: in cupy.core.core.compile_with_cache
    return cuda.compile_with_cache(
cupy/cuda/compiler.py:412: in compile_with_cache
    name_expressions, log_stream, cache_in_memory)
cupy/cuda/compiler.py:796: in _compile_with_cache_hip
    log_stream, cache_in_memory)
cupy/cuda/compiler.py:254: in compile_using_nvrtc
    name_expressions, log_stream, jitify)
cupy/cuda/compiler.py:237: in _compile
    ptx, mapping = prog.compile(options, log_stream)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <cupy.cuda.compiler._NVRTCProgram object at 0x7f4b9ea4d750>
options = ('-I/home/leofang/dev/cupy_rocm350/cupy/core/include', '-I/opt/rocm/include', '-arch=gfx906'), log_stream = None

    def compile(self, options=(), log_stream=None):
        try:
            if self.name_expressions:
                for ker in self.name_expressions:
                    nvrtc.addAddNameExpression(self.ptr, ker)
            nvrtc.compileProgram(self.ptr, options)
            mapping = None
            if self.name_expressions:
                mapping = {}
                for ker in self.name_expressions:
                    mapping[ker] = nvrtc.getLoweredName(self.ptr, ker)
            if log_stream is not None:
                log_stream.write(nvrtc.getProgramLog(self.ptr))
            # TODO(leofang): use getCUBIN() for _cuda_version >= 11010?
            return nvrtc.getPTX(self.ptr), mapping
        except nvrtc.NVRTCError:
            log = nvrtc.getProgramLog(self.ptr)
            raise CompileException(log, self.src, self.name, options,
>                                  'nvrtc' if not runtime.is_hip else 'hiprtc')
E           cupy.cuda.compiler.CompileException: /tmp/comgr-dbc4ec/input/CompileSource:963:7: error: no matching function for call to 'abs'
E             if (abs(z.real()) < ::sqrtf(FLT_MIN) && abs(z.imag()) < ::sqrtf(FLT_MIN)) {
E                 ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:911:30: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'float'
E           __host__ __device__ inline T abs(const complex<T>& z) {
E                                        ^
E           /tmp/comgr-dbc4ec/input/CompileSource:963:31: error: use of undeclared identifier 'FLT_MIN'
E             if (abs(z.real()) < ::sqrtf(FLT_MIN) && abs(z.imag()) < ::sqrtf(FLT_MIN)) {
E                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:963:43: error: no matching function for call to 'abs'
E             if (abs(z.real()) < ::sqrtf(FLT_MIN) && abs(z.imag()) < ::sqrtf(FLT_MIN)) {
E                                                     ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:911:30: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'float'
E           __host__ __device__ inline T abs(const complex<T>& z) {
E                                        ^
E           /tmp/comgr-dbc4ec/input/CompileSource:963:67: error: use of undeclared identifier 'FLT_MIN'
E             if (abs(z.real()) < ::sqrtf(FLT_MIN) && abs(z.imag()) < ::sqrtf(FLT_MIN)) {
E                                                                             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:973:7: error: no matching function for call to 'abs'
E             if (abs(z.real()) < ::sqrt(DBL_MIN) && abs(z.imag()) < ::sqrt(DBL_MIN)) {
E                 ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:911:30: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline T abs(const complex<T>& z) {
E                                        ^
E           /tmp/comgr-dbc4ec/input/CompileSource:973:30: error: use of undeclared identifier 'DBL_MIN'
E             if (abs(z.real()) < ::sqrt(DBL_MIN) && abs(z.imag()) < ::sqrt(DBL_MIN)) {
E                                        ^
E           /tmp/comgr-dbc4ec/input/CompileSource:973:42: error: no matching function for call to 'abs'
E             if (abs(z.real()) < ::sqrt(DBL_MIN) && abs(z.imag()) < ::sqrt(DBL_MIN)) {
E                                                    ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:911:30: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline T abs(const complex<T>& z) {
E                                        ^
E           /tmp/comgr-dbc4ec/input/CompileSource:973:65: error: use of undeclared identifier 'DBL_MIN'
E             if (abs(z.real()) < ::sqrt(DBL_MIN) && abs(z.imag()) < ::sqrt(DBL_MIN)) {
E                                                                           ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1164:23: error: use of undeclared identifier 'DBL_EPSILON'
E               if (y == 1 && x < DBL_EPSILON * DBL_EPSILON / 128) {
E                                 ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1164:37: error: use of undeclared identifier 'DBL_EPSILON'
E               if (y == 1 && x < DBL_EPSILON * DBL_EPSILON / 128) {
E                                               ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1169:13: error: no matching function for call to 'sqrt'
E                 *rx = sqrt(x);
E                       ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:507:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sqrt(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1170:21: error: use of undeclared identifier 'DBL_EPSILON'
E               } else if (x >= DBL_EPSILON * fabs(y - 1)) {
E                               ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1176:25: error: no matching function for call to 'sqrt'
E                 *rx = log1p(Am1 + sqrt(Am1 * (A + 1)));
E                                   ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:507:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sqrt(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1182:17: error: no matching function for call to 'sqrt'
E                 *rx = x / sqrt((1 - y) * (1 + y));
E                           ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:507:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sqrt(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1187:29: error: no matching function for call to 'sqrt'
E                 *rx = log1p((y - 1) + sqrt((y - 1) * (y + 1)));
E                                       ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:507:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sqrt(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1190:19: error: no matching function for call to 'sqrt'
E               *rx = log(A + sqrt(A * A - 1));
E                             ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:507:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sqrt(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1202:28: error: use of undeclared identifier 'DBL_EPSILON'
E               *sqrt_A2my2 = A * (2 / DBL_EPSILON);
E                                      ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1203:23: error: use of undeclared identifier 'DBL_EPSILON'
E               *new_y = y * (2 / DBL_EPSILON);
E                                 ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1217:23: error: use of undeclared identifier 'DBL_EPSILON'
E               if (y == 1 && x < DBL_EPSILON / 128) {
E                                 ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1222:21: error: no matching function for call to 'sqrt'
E                 *sqrt_A2my2 = sqrt(x) * sqrt((A + y) / 2);
E                               ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:507:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sqrt(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1222:31: error: no matching function for call to 'sqrt'
E                 *sqrt_A2my2 = sqrt(x) * sqrt((A + y) / 2);
E                                         ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:507:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sqrt(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1223:21: error: use of undeclared identifier 'DBL_EPSILON'
E               } else if (x >= DBL_EPSILON * fabs(y - 1)) {
E                               ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1231:21: error: no matching function for call to 'sqrt'
E                 *sqrt_A2my2 = sqrt(Amy * (A + y));
E                               ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:507:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sqrt(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1241:53: error: no matching function for call to 'sqrt'
E                     x * (4 / DBL_EPSILON / DBL_EPSILON) * y / sqrt((y + 1) * (y - 1));
E                                                               ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:507:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sqrt(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1241:20: error: use of undeclared identifier 'DBL_EPSILON'
E                     x * (4 / DBL_EPSILON / DBL_EPSILON) * y / sqrt((y + 1) * (y - 1));
E                              ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1241:34: error: use of undeclared identifier 'DBL_EPSILON'
E                     x * (4 / DBL_EPSILON / DBL_EPSILON) * y / sqrt((y + 1) * (y - 1));
E                                            ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1242:25: error: use of undeclared identifier 'DBL_EPSILON'
E                 *new_y = y * (4 / DBL_EPSILON / DBL_EPSILON);
E                                   ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1242:39: error: use of undeclared identifier 'DBL_EPSILON'
E                 *new_y = y * (4 / DBL_EPSILON / DBL_EPSILON);
E                                                 ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1248:21: error: no matching function for call to 'sqrt'
E                 *sqrt_A2my2 = sqrt((1 - y) * (1 + y));
E                               ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:507:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sqrt(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1265:38: error: use of undeclared identifier 'DBL_EPSILON'
E             const double RECIP_EPSILON = 1.0 / DBL_EPSILON;
E                                                ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1308:10: error: no matching function for call to 'asin'
E               ry = asin(B);
E                    ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:575:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> asin(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1355:51: error: use of undeclared identifier 'infinity'
E               if (isinf(x)) return (complex<double>(y + y, -infinity<double>()));
E                                                             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1368:38: error: use of undeclared identifier 'DBL_EPSILON'
E             const double RECIP_EPSILON = 1.0 / DBL_EPSILON;
E                                                ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1392:12: error: no matching function for call to 'acos'
E                 rx = acos(B);
E                      ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:565:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> acos(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1394:12: error: no matching function for call to 'acos'
E                 rx = acos(-B);
E                      ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:565:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> acos(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1451:12: error: use of undeclared identifier 'DBL_MAX'
E             if (ax > DBL_MAX / 2)
E                      ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1452:29: error: no matching function for call to 'log'
E               return (complex<double>(log(hypot(x / m_e, y / m_e)) + 1, atan2(y, x)));
E                                       ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:431:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> log(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1463:29: error: no matching function for call to 'log'
E               return (complex<double>(log(hypot(x, y)), atan2(y, x)));
E                                       ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:431:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> log(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1465:27: error: no matching function for call to 'log'
E             return (complex<double>(log(ax * ax + ay * ay) / 2, atan2(y, x)));
E                                     ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:431:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> log(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1508:3: error: use of undeclared identifier 'get_high_word'
E             get_high_word(hx, x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1510:3: error: use of undeclared identifier 'get_high_word'
E             get_high_word(hy, y);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1513:20: error: use of undeclared identifier 'DBL_MAX_EXP'
E             const int BIAS = DBL_MAX_EXP - 1;
E                              ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1516:23: error: use of undeclared identifier 'DBL_MANT_DIG'
E             const int CUTOFF = (DBL_MANT_DIG / 2 + 1);
E                                 ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1521:21: error: use of undeclared identifier 'DBL_MAX_EXP'
E             if (ix <= (BIAS + DBL_MAX_EXP / 2 - CUTOFF) << 20)
E                               ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1524:3: error: use of undeclared identifier 'set_high_word'
E             set_high_word(scale, 0x7ff00000 - ix); /* 2**(1-ilogb(x)) */
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1555:50: error: no matching function for call to 'atanh'
E             if (y == 0 && ax <= 1) return (complex<double>(atanh(x), y));
E                                                            ^~~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:617:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> atanh(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1558:42: error: no matching function for call to 'atan'
E             if (x == 0) return (complex<double>(x, atan(y)));
E                                                    ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:585:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> atan(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1575:38: error: use of undeclared identifier 'DBL_EPSILON'
E             const double RECIP_EPSILON = 1.0 / DBL_EPSILON;
E                                                ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1593:23: error: use of undeclared identifier 'DBL_EPSILON'
E             if (ax == 1 && ay < DBL_EPSILON)
E                                 ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1594:19: error: no matching function for call to 'log'
E               rx = (m_ln2 - log(ay)) / 2;
E                             ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:431:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> log(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1600:17: error: use of undeclared identifier 'DBL_EPSILON'
E             else if (ay < DBL_EPSILON)
E                           ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1822:23: error: use of undeclared identifier 'FLT_EPSILON'
E               if (y == 1 && x < FLT_EPSILON * FLT_EPSILON / 128) {
E                                 ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1822:37: error: use of undeclared identifier 'FLT_EPSILON'
E               if (y == 1 && x < FLT_EPSILON * FLT_EPSILON / 128) {
E                                               ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1824:21: error: use of undeclared identifier 'FLT_EPSILON'
E               } else if (x >= FLT_EPSILON * fabsf(y - 1)) {
E                               ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1840:28: error: use of undeclared identifier 'FLT_EPSILON'
E               *sqrt_A2my2 = A * (2 / FLT_EPSILON);
E                                      ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1841:23: error: use of undeclared identifier 'FLT_EPSILON'
E               *new_y = y * (2 / FLT_EPSILON);
E                                 ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1850:23: error: use of undeclared identifier 'FLT_EPSILON'
E               if (y == 1 && x < FLT_EPSILON / 128) {
E                                 ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1852:21: error: use of undeclared identifier 'FLT_EPSILON'
E               } else if (x >= FLT_EPSILON * fabsf(y - 1)) {
E                               ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1857:20: error: use of undeclared identifier 'FLT_EPSILON'
E                     x * (4 / FLT_EPSILON / FLT_EPSILON) * y / sqrtf((y + 1) * (y - 1));
E                              ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1857:34: error: use of undeclared identifier 'FLT_EPSILON'
E                     x * (4 / FLT_EPSILON / FLT_EPSILON) * y / sqrtf((y + 1) * (y - 1));
E                                            ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1858:25: error: use of undeclared identifier 'FLT_EPSILON'
E                 *new_y = y * (4 / FLT_EPSILON / FLT_EPSILON);
E                                   ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1858:39: error: use of undeclared identifier 'FLT_EPSILON'
E                 *new_y = y * (4 / FLT_EPSILON / FLT_EPSILON);
E                                                 ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1869:37: error: use of undeclared identifier 'FLT_EPSILON'
E             const float RECIP_EPSILON = 1.0 / FLT_EPSILON;
E                                               ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1930:50: error: use of undeclared identifier 'infinity'
E               if (isinf(x)) return (complex<float>(y + y, -infinity<float>()));
E                                                            ^
E           /tmp/comgr-dbc4ec/input/CompileSource:1936:37: error: use of undeclared identifier 'FLT_EPSILON'
E             const float RECIP_EPSILON = 1.0 / FLT_EPSILON;
E                                               ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2004:12: error: use of undeclared identifier 'FLT_MAX'
E             if (ax > FLT_MAX / 2)
E                      ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2044:3: error: use of undeclared identifier 'get_float_word'
E             get_float_word(hx, x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2046:3: error: use of undeclared identifier 'get_float_word'
E             get_float_word(hy, y);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2049:20: error: use of undeclared identifier 'FLT_MAX_EXP'
E             const int BIAS = FLT_MAX_EXP - 1;
E                              ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2051:23: error: use of undeclared identifier 'FLT_MANT_DIG'
E             const int CUTOFF = (FLT_MANT_DIG / 2 + 1);
E                                 ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2054:21: error: use of undeclared identifier 'FLT_MAX_EXP'
E             if (ix <= (BIAS + FLT_MAX_EXP / 2 - CUTOFF) << 23)
E                               ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2056:3: error: use of undeclared identifier 'set_float_word'
E             set_float_word(scale, 0x7f800000 - ix);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2085:38: error: use of undeclared identifier 'FLT_EPSILON'
E             const float RECIP_EPSILON = 1.0f / FLT_EPSILON;
E                                                ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2097:23: error: use of undeclared identifier 'FLT_EPSILON'
E             if (ax == 1 && ay < FLT_EPSILON)
E                                 ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2104:17: error: use of undeclared identifier 'FLT_EPSILON'
E             else if (ay < FLT_EPSILON)
E                           ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2230:3: error: use of undeclared identifier 'extract_words'
E             extract_words(hx, lx, x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2231:3: error: use of undeclared identifier 'extract_words'
E             extract_words(hy, ly, y);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2247:43: error: no matching function for call to 'cos'
E                 return (thrust::complex<double>(h * cos(y), copysign(h, x) * sin(y)));
E                                                     ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:516:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> cos(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2247:68: error: no matching function for call to 'sin'
E                 return (thrust::complex<double>(h * cos(y), copysign(h, x) * sin(y)));
E                                                                              ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:523:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sin(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2251:12: error: use of undeclared identifier 'ldexp_cexp'
E                 z_ = ldexp_cexp(thrust::complex<double>(fabs(x), y), -1);
E                      ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2256:47: error: no matching function for call to 'cos'
E                 return (thrust::complex<double>(h * h * cos(y), h * sin(y)));
E                                                         ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:516:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> cos(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2256:59: error: no matching function for call to 'sin'
E                 return (thrust::complex<double>(h * h * cos(y), h * sin(y)));
E                                                                     ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:523:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sin(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2306:47: error: no matching function for call to 'cos'
E               return (thrust::complex<double>((x * x) * cos(y), x * sin(y)));
E                                                         ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:516:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> cos(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2306:59: error: no matching function for call to 'sin'
E               return (thrust::complex<double>((x * x) * cos(y), x * sin(y)));
E                                                                     ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:523:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sin(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2425:3: error: use of undeclared identifier 'get_float_word'
E             get_float_word(hx, x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2426:3: error: use of undeclared identifier 'get_float_word'
E             get_float_word(hy, y);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2445:12: error: use of undeclared identifier 'ldexp_cexpf'
E                 z_ = ldexp_cexpf(complex<float>(fabsf(x), y), -1);
E                      ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2569:11: error: no matching function for call to 'exp'
E             exp_x = exp(x - kln2);
E                     ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:424:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> exp(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2570:3: error: use of undeclared identifier 'get_high_word'
E             get_high_word(hx, exp_x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2572:3: error: use of undeclared identifier 'set_high_word'
E             set_high_word(exp_x, (hx & 0xfffff) | ((0x3ff + 1023) << 20));
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2590:3: error: use of undeclared identifier 'insert_words'
E             insert_words(scale1, (0x3ff + half_expt) << 20, 0);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2592:3: error: use of undeclared identifier 'insert_words'
E             insert_words(scale2, (0x3ff + half_expt) << 20, 0);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2608:3: error: use of undeclared identifier 'extract_words'
E             extract_words(hy, ly, y);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2612:47: error: no matching function for call to 'exp'
E             if ((hy | ly) == 0) return (complex<double>(exp(x), y));
E                                                         ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:424:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> exp(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2613:3: error: use of undeclared identifier 'extract_words'
E             extract_words(hx, lx, x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2615:62: error: no matching function for call to 'cos'
E             if (((hx & 0x7fffffff) | lx) == 0) return (complex<double>(cos(y), sin(y)));
E                                                                        ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2334:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> cos(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2615:70: error: no matching function for call to 'sin'
E             if (((hx & 0x7fffffff) | lx) == 0) return (complex<double>(cos(y), sin(y)));
E                                                                                ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:523:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sin(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2645:37: error: no matching function for call to 'cos'
E               return (complex<double>(exp_x * cos(y), exp_x * sin(y)));
E                                               ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2334:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> cos(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2645:53: error: no matching function for call to 'sin'
E               return (complex<double>(exp_x * cos(y), exp_x * sin(y)));
E                                                               ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:523:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sin(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2729:3: error: use of undeclared identifier 'get_float_word'
E             get_float_word(hx, exp_x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2731:3: error: use of undeclared identifier 'set_float_word'
E             set_float_word(exp_x, (hx & 0x7fffff) | ((0x7f + 127) << 23));
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2745:3: error: use of undeclared identifier 'set_float_word'
E             set_float_word(scale1, (0x7f + half_expt) << 23);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2747:3: error: use of undeclared identifier 'set_float_word'
E             set_float_word(scale2, (0x7f + half_expt) << 23);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2749:26: error: no matching function for call to 'cos'
E             return (complex<float>(cos(y) * exp_x * scale1 * scale2,
E                                    ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2334:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'float'
E           __host__ __device__ inline complex<ValueType> cos(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2750:26: error: no matching function for call to 'sin'
E                                    sin(y) * exp_x * scale1 * scale2));
E                                    ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:523:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'float'
E           __host__ __device__ complex<T> sin(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2763:3: error: use of undeclared identifier 'get_float_word'
E             get_float_word(hy, y);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2767:39: error: no matching function for call to 'exp'
E             if (hy == 0) return (complex<float>(exp(x), y));
E                                                 ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2654:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'float'
E           __host__ __device__ inline complex<ValueType> exp(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2768:3: error: use of undeclared identifier 'get_float_word'
E             get_float_word(hx, x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2771:28: error: no matching function for call to 'cos'
E               return (complex<float>(cos(y), sin(y)));
E                                      ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2334:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'float'
E           __host__ __device__ inline complex<ValueType> cos(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2771:36: error: no matching function for call to 'sin'
E               return (complex<float>(cos(y), sin(y)));
E                                              ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:523:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'float'
E           __host__ __device__ complex<T> sin(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2872:3: error: use of undeclared identifier 'get_high_word'
E             get_high_word(hi, x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:2873:3: error: use of undeclared identifier 'insert_words'
E             insert_words(x, hi & 0xfffffff8, 0);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3073:3: error: use of undeclared identifier 'get_float_word'
E             get_float_word(hx, x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3076:3: error: use of undeclared identifier 'set_float_word'
E             set_float_word(ret, hx);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3428:27: error: use of undeclared identifier 'infinity'
E               return complex<float>(infinity<float>(), copysignf(0.0, z.imag()));
E                                     ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3437:28: error: use of undeclared identifier 'infinity'
E               return complex<double>(infinity<double>(), copysign(0.0, z.imag()));
E                                      ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3523:3: error: use of undeclared identifier 'extract_words'
E             extract_words(hx, lx, x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3524:3: error: use of undeclared identifier 'extract_words'
E             extract_words(hy, ly, y);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3531:49: error: no matching function for call to 'sinh'
E               if ((iy | ly) == 0) return (complex<double>(sinh(x), y));
E                                                           ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:546:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sinh(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3533:31: error: no matching function for call to 'sinh'
E                 return (complex<double>(sinh(x) * cos(y), cosh(x) * sin(y)));
E                                         ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:546:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sinh(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3533:41: error: no matching function for call to 'cos'
E                 return (complex<double>(sinh(x) * cos(y), cosh(x) * sin(y)));
E                                                   ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2334:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> cos(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3533:49: error: no matching function for call to 'cosh'
E                 return (complex<double>(sinh(x) * cos(y), cosh(x) * sin(y)));
E                                                           ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2341:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> cosh(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3533:59: error: no matching function for call to 'sin'
E                 return (complex<double>(sinh(x) * cos(y), cosh(x) * sin(y)));
E                                                                     ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:523:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sin(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3538:11: error: no matching function for call to 'exp'
E                 h = exp(fabs(x)) * 0.5;
E                     ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2654:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> exp(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3539:48: error: no matching function for call to 'cos'
E                 return (complex<double>(copysign(h, x) * cos(y), h * sin(y)));
E                                                          ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2334:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> cos(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3539:60: error: no matching function for call to 'sin'
E                 return (complex<double>(copysign(h, x) * cos(y), h * sin(y)));
E                                                                      ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:523:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sin(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3547:35: error: no matching function for call to 'cos'
E                 return (complex<double>(h * cos(y), h * h * sin(y)));
E                                             ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2334:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> cos(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3547:51: error: no matching function for call to 'sin'
E                 return (complex<double>(h * cos(y), h * h * sin(y)));
E                                                             ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:523:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sin(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3597:33: error: no matching function for call to 'cos'
E               return (complex<double>(x * cos(y), infinity<double>() * sin(y)));
E                                           ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2334:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> cos(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3597:41: error: use of undeclared identifier 'infinity'
E               return (complex<double>(x * cos(y), infinity<double>() * sin(y)));
E                                                   ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3597:62: error: no matching function for call to 'sin'
E               return (complex<double>(x * cos(y), infinity<double>() * sin(y)));
E                                                                        ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:523:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sin(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3714:3: error: use of undeclared identifier 'get_float_word'
E             get_float_word(hx, x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3715:3: error: use of undeclared identifier 'get_float_word'
E             get_float_word(hy, y);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3754:41: error: use of undeclared identifier 'infinity'
E               return (complex<float>(x * cosf(y), infinity<float>() * sinf(y)));
E                                                   ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3851:41: error: use of undeclared identifier 'infinity'
E             if (isinf(b)) return (complex<double>(infinity<double>(), b));
E                                                   ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3891:9: error: no matching function for call to 'sqrt'
E               t = sqrt((a + hypot(a, b)) * 0.5);
E                   ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:507:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sqrt(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3894:9: error: no matching function for call to 'sqrt'
E               t = sqrt((-a + hypot(a, b)) * 0.5);
E                   ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:507:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> sqrt(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:3990:40: error: use of undeclared identifier 'infinity'
E             if (isinf(b)) return (complex<float>(infinity<float>(), b));
E                                                  ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4165:3: error: use of undeclared identifier 'extract_words'
E             extract_words(hx, lx, x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4187:5: error: use of undeclared identifier 'set_high_word'
E               set_high_word(x, hx - 0x40000000); /* x = copysign(1, x) */
E               ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4188:61: error: no matching function for call to 'sin'
E               return (complex<double>(x, copysign(0.0, isinf(y) ? y : sin(y) * cos(y))));
E                                                                       ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:3625:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> sin(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4188:70: error: no matching function for call to 'cos'
E               return (complex<double>(x, copysign(0.0, isinf(y) ? y : sin(y) * cos(y))));
E                                                                                ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2334:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> cos(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4203:21: error: no matching function for call to 'exp'
E               double exp_mx = exp(-fabs(x));
E                               ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2654:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> exp(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4205:35: error: no matching function for call to 'sin'
E                                       4.0 * sin(y) * cos(y) * exp_mx * exp_mx));
E                                             ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:3625:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> sin(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4205:44: error: no matching function for call to 'cos'
E                                       4.0 * sin(y) * cos(y) * exp_mx * exp_mx));
E                                                      ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:2334:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> cos(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4209:7: error: no matching function for call to 'tan'
E             t = tan(y);
E                 ^~~
E           /tmp/comgr-dbc4ec/input/CompileSource:530:32: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ complex<T> tan(const complex<T>& z);
E                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4211:7: error: no matching function for call to 'sinh'
E             s = sinh(x);
E                 ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:3632:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> sinh(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4212:9: error: no matching function for call to 'sqrt'
E             rho = sqrt(1.0 + s * s); /* = cosh(x) */
E                   ^~~~
E           /tmp/comgr-dbc4ec/input/CompileSource:3912:47: note: candidate template ignored: could not match 'complex<type-parameter-0-0>' against 'double'
E           __host__ __device__ inline complex<ValueType> sqrt(const complex<ValueType>& z) {
E                                                         ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4319:3: error: use of undeclared identifier 'get_float_word'
E             get_float_word(hx, x);
E             ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4324:5: error: use of undeclared identifier 'set_float_word'
E               set_float_word(x, hx - 0x40000000);
E               ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4569:12: error: unknown type name 'float16'
E           __device__ float16 atomicAdd(float16* address, float16 val) {
E                      ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4569:30: error: unknown type name 'float16'
E           __device__ float16 atomicAdd(float16* address, float16 val) {
E                                        ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4569:48: error: unknown type name 'float16'
E           __device__ float16 atomicAdd(float16* address, float16 val) {
E                                                          ^
E           /tmp/comgr-dbc4ec/input/CompileSource:4589:10: error: use of undeclared identifier 'float16'
E             return float16(raw);
E                    ^
E           154 errors generated when compiling for gfx906.
E           Error: Failed to compile opencl source (from CL or HIP source to LLVM IR).

cupy/cuda/compiler.py:618: CompileException
======================================================================= short test summary info =======================================================================
FAILED tests/cupy_tests/core_tests/test_raw.py::TestRaw_param_0_{backend='nvrtc', in_memory=False}::test_basic - cupy.cuda.compiler.CompileException: /tmp/comgr-dbc...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
========================================================================== 1 failed in 1.90s ==========================================================================
```

Reply:
I wonder why ...

Reply:
So it's because `_cupy_extra_header_list` is sorted. Reverted to the original order resolves many (but not all) errors.